### ISSUE TITLE
[Question] Caprover with custom certificate on company VM behind firewall

### ISSUE THREAD

**Issue by JosXa** [caprover/caprover#887]
I enjoy using Caprover for my side projects on my public domain and figured I'd also run it at work to simplify managing some internal tools and make use of the one-click apps. Turns out our IT has very strict firewall settings and I've had to request quite a few outgoing domains to be opened for the VM I'm trying to run it on, however I eventually managed to properly install Captain and have it be accessible via VPN. I was provided with a certificate and added that to the nginx configurations manually, which also worked out after a few attempts. 

As the _Enable SSL_ button wasn't working, in order to check that the cert was working alright, I also removed the condition from the following nginx declaration...
```
    <%
        if (captain.hasRootSsl) {
    %>
        listen              443 ssl;
        ssl_certificate     <%-captain.crtPath%>;
        ssl_certificate_key <%-captain.keyPath%>;

    <%
        }
    %>
```
... and updated the cert paths accordingly, upon which the dashboard was accessible via https ✔️:

```
        listen              443 ssl;
        ssl_certificate     /nginx-shared/mycert.crt;
        ssl_certificate_key /nginx-shared/mycert.rsa;
```

If I understand it right however, the `captain.hasRootSsl` variable is set to true when the certbot has been fed with the `register` command and an email address, which only happens when clicking the _Enable SSL_ button. This part seems to get blocked by the firewall however:

```
An unexpected error occurred:
captain-captain.1.1o01r4gu47y3@2219400-SERVERNAME    | requests.exceptions.SSLError: HTTPSConnectionPool(host='acme-v02.api.letsencrypt.org', port=443): Max retries exceeded with url: /directory (Caused by SSLError(SSLError("bad handshake: SysCallError(104, 'ECONNRESET')")))
``` 

I do think it will be alright to open the VM up to allow access to the letsencrypt API and I will request that ASAP, but as I'm not an expert in networking I'd just like to understand what the *best practices* are when running a setup like I'm describing here.

**1)** I was under the impression that a root wildcard certificate would already be enough to run **all** apps inside Caprover in this company-internal setup and am confused why it is required to use certbot to create a new cert for every single application. As editing the nginx root config file already worked out fine , I assume I could skip the Enable SSL button for all the apps aswell and essentially circumvent using certbot this way... 
**2)** Would there be any benefit using per-app-certificates issued by letsencrypt for company-internal resources or would that even be a security issue (for a security-conscious company like mine)? 
**3)** Do you see a way for me to enable SSL in caprover or at least have it recognize that SSL is technically enabled using either this "one cert for all" setup or maybe some special configuration for the certbot?

Thanks in advance :)

**Comment by JosXa** [comment]
> As editing the nginx root config file already worked out fine, I assume I could skip the Enable SSL button for all the apps aswell and essentially circumvent using certbot this way...

Update: Nope, always seeing the error page if I follow the same steps as above, i.e. "manually enable SSL by editing the app's nginx config".

**Comment by githubsaturn** [comment]
TLDR: CapRover is not really designed for these custom requirements. You'd have to do **a lot of manual fiddling** to make this work. 

> 1) I was under the impression that a root wildcard certificate would already be enough to run all apps inside Caprover in this company-internal setup and am confused why it is required to use certbot to create a new cert for every single application. As editing the nginx root config file already worked out fine , I assume I could skip the Enable SSL button for all the apps aswell and essentially circumvent using certbot this way...

Enable SSL button is tied to Certbot. You can use your own certificate, but then you are also responsible for "enabling SSL" by removing `if (captain.hasRootSsl)` as similar lines in each application. 

> 2) Would there be any benefit using per-app-certificates issued by letsencrypt for company-internal resources or would that even be a security issue (for a security-conscious company like mine)?

LetsEncrypt certs are valid pubic certs. So they are completely secure by the industry standard. Your company's internal certs could be insecure if they are self signed. If they are valid certs, they are also as secure as LetsEncrypt certs.


> 3) Do you see a way for me to enable SSL in caprover or at least have it recognize that SSL is technically enabled using either this "one cert for all" setup or maybe some special configuration for the certbot?

It won't be shown on the UI. But by the virtue of editing nginx config, you're essentially enabling SSL for your root domain and your apps.

**Comment by githubsaturn** [comment]
Thinking about this again, maybe this approach works even better:

```


                               VM1                     VM2
                         +------------+          +-------------+
                         |            |          |             |
                         |    Nginx   |          |  CapRover   |
   +--------+   HTTPS    |     or     |   HTTP   |   System    |
   |  User  +----------->+   other    +--------->+ (nginx+     |
   +--------+            |    load    |          | caprover+   |
                         |  balancer  |          | apps)       |
                         | with certs |          |             |
                         |            |          |             |
                         +------------+          +-------------+
```

**Comment by JosXa** [comment]
Ah, so you're suggesting to keep VM2 unencrypted and basically SSL offload on VM1 in that diagram?

What I'm still curious about is why you name adding a custom certificate to the nginx configuration as an example for use cases of `nginx-shared` when that approach is extremely hacky in practice. Is that typically only meant for individual apps and not for the base configuration?

**Comment by githubsaturn** [comment]
> Ah, so you're suggesting to keep VM2 unencrypted and basically SSL offload on VM1 in that diagram?

Yes, that means you won't be able to click on enabled HTTPS on your CapRover instance, but on VM1, you can have a wildcard cert with some nice HSTS that automatically redirects all HTTP requests to https and CapRover only deals with plain HTTP. 

> Is that typically only meant for individual apps and not for the base configuration?

Yes, but in general, `nginx-shared` directory is supposed to work for these custom configurations. I wouldn't call this "hacky". The only downside with that approach is that you have to manually configure individual nginx configuration for each app when you're creating a new app.

**Comment by JosXa** [comment]
Referenced #977 because of:

> Yes, but in general, nginx-shared directory is supposed to work for these custom configurations. I wouldn't call this "hacky". The only downside with that approach is that you have to manually configure individual nginx configuration for each app when you're creating a new app.
